File: big_bertha_hardware_bringup/big_bertha_hardware_bringup/hardware_bridge_node.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy
from sensor_msgs.msg import Imu
from std_msgs.msg import Float64MultiArray
from std_srvs.srv import Trigger

try:
    from arduino_router_bridge import Bridge
except ImportError:
    Bridge = None


class HardwareBridgeNode(Node):
    def __init__(self):
        super().__init__("hardware_bridge")

        self.declare_parameters(
            namespace="",
            parameters=[
                ("imu_frame_id", "imu_link"),
                ("imu_rate", 200.0),
                ("servo_count", 12),
                ("servo_freq", 50),
                ("servo_angle_min", -1.57),
                ("servo_angle_max", 1.57),
                ("servo_pulse_min", 500),
                ("servo_pulse_max", 2500),
            ],
        )

        self._imu_frame = self.get_parameter("imu_frame_id").value
        self._imu_rate = self.get_parameter("imu_rate").value

        self._servo_count = self.get_parameter("servo_count").value
        self._servo_freq = self.get_parameter("servo_freq").value
        self._angle_min = self.get_parameter("servo_angle_min").value
        self._angle_max = self.get_parameter("servo_angle_max").value
        self._pulse_min = self.get_parameter("servo_pulse_min").value
        self._pulse_max = self.get_parameter("servo_pulse_max").value

        self._bridge = None
        self._init_bridge()

        imu_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            depth=1,
        )

        self._imu_pub = self.create_publisher(Imu, "/imu", imu_qos)
        self._cmd_sub = self.create_subscription(
            Float64MultiArray,
            "/position_controller/commands",
            self._cmd_callback,
            1,
        )
        self._ping_srv = self.create_service(
            Trigger, "/hardware/ping", self._ping_callback
        )

        period = 1.0 / self._imu_rate
        self._imu_timer = self.create_timer(period, self._publish_imu)

        self.get_logger().info(
            f"Started | IMU @ {self._imu_rate} Hz on /{self._imu_frame}"
            f" | {self._servo_count} servos via Bridge"
        )

    def _init_bridge(self):
        if Bridge is None:
            self.get_logger().warning(
                "arduino_router_bridge not found — running in mock mode"
            )
            self._bridge = MockBridge()
            return
        try:
            self._bridge = Bridge()
            self.get_logger().info("Arduino Router Bridge connected")
        except Exception as e:
            self.get_logger().error(f"Bridge init failed: {e}")
            self._bridge = MockBridge()

    def _publish_imu(self):
        if self._bridge is None:
            return
        try:
            raw = self._bridge.call("read_imu")
        except Exception as e:
            self.get_logger().warn(f"read_imu failed: {e}", throttle_duration=5.0)
            return

        parts = raw.strip().split()
        if len(parts) < 6:
            return

        gx = float(parts[0])
        gy = float(parts[1])
        gz = float(parts[2])
        ax = float(parts[3])
        ay = float(parts[4])
        az = float(parts[5])

        msg = Imu()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = self._imu_frame

        msg.angular_velocity.x = gx
        msg.angular_velocity.y = gy
        msg.angular_velocity.z = gz

        msg.linear_acceleration.x = ax
        msg.linear_acceleration.y = ay
        msg.linear_acceleration.z = az

        self._imu_pub.publish(msg)

    def _cmd_callback(self, msg: Float64MultiArray):
        if self._bridge is None:
            return
        if len(msg.data) < self._servo_count:
            self.get_logger().warn(
                f"cmd has {len(msg.data)} values, expected {self._servo_count}"
            )
            return

        rads = msg.data[: self._servo_count]
        pulse_str = " ".join(str(self._rad_to_pulse(r)) for r in rads)

        try:
            self._bridge.call("set_servos", pulse_str)
        except Exception as e:
            self.get_logger().warn(f"set_servos failed: {e}", throttle_duration=2.0)

    def _rad_to_pulse(self, rad: float) -> int:
        t = (rad - self._angle_min) / (self._angle_max - self._angle_min)
        t = max(0.0, min(1.0, t))
        return int(self._pulse_min + t * (self._pulse_max - self._pulse_min))

    def _ping_callback(self, request, response):
        if self._bridge is None:
            response.success = False
            response.message = "bridge not available"
            return response
        try:
            result = self._bridge.call("ping")
            response.success = result.strip() == "ok"
            response.message = result.strip()
        except Exception as e:
            response.success = False
            response.message = str(e)
        return response


class MockBridge:
    def call(self, fn, *args):
        if fn == "read_imu":
            return "0.0 0.0 0.0 0.0 0.0 9.81"
        return "ok"


def main(args=None):
    rclpy.init(args=args)
    node = HardwareBridgeNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: big_bertha_hardware_bringup/src/hardware_bridge_node.cpp
#include <algorithm>
#include <cmath>
#include <memory>
#include <string>
#include <sstream>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/imu.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include "std_srvs/srv/trigger.hpp"

#include "big_bertha_hardware_bringup/router_bridge.hpp"

class HardwareBridgeNode : public rclcpp::Node {
public:
  HardwareBridgeNode()
  : Node("hardware_bridge"), bridge_(nullptr) {
    declare_parameter("imu_frame_id", "imu_link");
    declare_parameter("imu_rate", 30.0);
    declare_parameter("servo_count", 12);
    declare_parameter("servo_freq", 50);
    declare_parameter("servo_angle_min", -1.57);
    declare_parameter("servo_angle_max", 1.57);
    declare_parameter("servo_pulse_min", 500);
    declare_parameter("servo_pulse_max", 2500);
    declare_parameter("router_socket", "/var/run/arduino-router.sock");
    declare_parameter("mock", false);

    imu_frame_ = get_parameter("imu_frame_id").as_string();
    imu_rate_ = get_parameter("imu_rate").as_double();
    servo_count_ = get_parameter("servo_count").as_int();
    angle_min_ = get_parameter("servo_angle_min").as_double();
    angle_max_ = get_parameter("servo_angle_max").as_double();
    pulse_min_ = get_parameter("servo_pulse_min").as_int();
    pulse_max_ = get_parameter("servo_pulse_max").as_int();
    bool mock = get_parameter("mock").as_bool();

    init_bridge(mock);

    imu_pub_ = create_publisher<sensor_msgs::msg::Imu>("/imu", 1);
    cmd_sub_ = create_subscription<std_msgs::msg::Float64MultiArray>(
        "/position_controller/commands", 1,
        std::bind(&HardwareBridgeNode::cmd_callback, this,
                  std::placeholders::_1));
    ping_srv_ = create_service<std_srvs::srv::Trigger>(
        "/hardware/ping",
        std::bind(&HardwareBridgeNode::ping_callback, this,
                  std::placeholders::_1, std::placeholders::_2));

    auto period = std::chrono::duration<double>(1.0 / imu_rate_);
    imu_timer_ = create_wall_timer(
        period, std::bind(&HardwareBridgeNode::publish_imu, this));

    RCLCPP_INFO(get_logger(),
                "Started | IMU @ %.1f Hz on /%s | %ld servos | %s",
                imu_rate_, imu_frame_.c_str(), servo_count_,
                mock ? "MOCK" : "HW");
  }

private:
  void init_bridge(bool mock) {
    if (mock) {
      RCLCPP_WARN(get_logger(), "Mock mode enabled — no hardware");
      bridge_ = std::make_unique<MockBridge>();
      return;
    }
    try {
      std::string sock_path = get_parameter("router_socket").as_string();
      bridge_ = std::make_unique<RouterBridge>(sock_path);
      RCLCPP_INFO(get_logger(), "Connected to router at %s",
                  sock_path.c_str());
    } catch (const std::exception & e) {
      RCLCPP_ERROR(get_logger(), "RouterBridge init failed: %s", e.what());
      RCLCPP_WARN(get_logger(), "Falling back to mock mode");
      bridge_ = std::make_unique<MockBridge>();
    }
  }

  void publish_imu() {
    if (!bridge_) return;
    std::string raw;
    try {
      raw = bridge_->call("read_imu");
    } catch (const std::exception & e) {
      RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 5000,
                           "read_imu failed: %s", e.what());
      return;
    }

    std::vector<double> parts;
    std::istringstream iss(raw);
    double val;
    while (iss >> val) {
      parts.push_back(val);
    }
    if (parts.size() < 6) return;

    auto msg = sensor_msgs::msg::Imu();
    msg.header.stamp = now();
    msg.header.frame_id = imu_frame_;
    msg.angular_velocity.x = parts[0];
    msg.angular_velocity.y = parts[1];
    msg.angular_velocity.z = parts[2];
    msg.linear_acceleration.x = parts[3];
    msg.linear_acceleration.y = parts[4];
    msg.linear_acceleration.z = parts[5];
    imu_pub_->publish(msg);
  }

  void cmd_callback(const std_msgs::msg::Float64MultiArray::SharedPtr msg) {
    if (!bridge_) return;
    if (static_cast<int64_t>(msg->data.size()) < servo_count_) {
      RCLCPP_WARN(get_logger(),
                  "cmd has %zu values, expected %ld",
                  msg->data.size(), servo_count_);
      return;
    }

    std::ostringstream pulse_str;
    for (int64_t i = 0; i < servo_count_; ++i) {
      if (i > 0) pulse_str << " ";
      pulse_str << rad_to_pulse(msg->data[i]);
    }

    try {
      bridge_->call("set_servos", pulse_str.str());
    } catch (const std::exception & e) {
      RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 2000,
                           "set_servos failed: %s", e.what());
    }
  }

  int rad_to_pulse(double rad) const {
    double t = (rad - angle_min_) / (angle_max_ - angle_min_);
    t = std::max(0.0, std::min(1.0, t));
    return static_cast<int>(pulse_min_ + t * (pulse_max_ - pulse_min_));
  }

  void ping_callback(
      const std::shared_ptr<std_srvs::srv::Trigger::Request> request,
      std::shared_ptr<std_srvs::srv::Trigger::Response> response) {
    (void)request;
    if (!bridge_) {
      response->success = false;
      response->message = "bridge not available";
      return;
    }
    try {
      std::string result = bridge_->call("ping");
      response->success = (result == "ok");
      response->message = result;
    } catch (const std::exception & e) {
      response->success = false;
      response->message = e.what();
    }
  }

  // -- members --
  std::unique_ptr<BridgeInterface> bridge_;
  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr imu_pub_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr cmd_sub_;
  rclcpp::Service<std_srvs::srv::Trigger>::SharedPtr ping_srv_;
  rclcpp::TimerBase::SharedPtr imu_timer_;

  std::string imu_frame_;
  double imu_rate_;
  int64_t servo_count_;
  double angle_min_;
  double angle_max_;
  int pulse_min_;
  int pulse_max_;
};

int main(int argc, char * argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<HardwareBridgeNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
